neural/neural_discern_run_script.py

```diff
@@ -0,0 +1,342 @@
+
+from neural.data_processor import DataDictProcessor
+from neural.dataset import generate_docpartition_per_question, validate_q_docpartitions, compute_class_weights_per_fold_
+from neural.model import BertEmbedder, generate_sents_embeds_from_docs
+from neural.run_workflow import generate_models_config, HyperparamConfig, hyperparam_model_search_parallel, \
+    get_best_config_from_hyperparamsearch, train_val_run, train_val_run_one_question, test_run
+from neural.utilities import ReaderWriter, create_directory
+
+
+# might have to be immediately after import torch.multiprocessing as mp
+mp.set_start_method("spawn", force=True)
+
+
+def read_pickles(data_dir):
+
+    # Read stored data structures
+    # {question:{fold_num:{dsettype:np.array(list of doc_ids)}}}
+    q_partitions = ReaderWriter.read_data(os.path.join(data_dir, 'q_partitions.pkl'))
+    # instance of :class:`DocDataTensor`
+    docs_data_tensor = ReaderWriter.read_data(os.path.join(data_dir, 'docs_data_tensor.pkl'))
+    # instance variable for processed articles dict
+    proc_articles_dict = ReaderWriter.read_data(os.path.join(data_dir, 'processor_articles_dict.pkl'))
+    # instance variable for computed articles reprs
+    proc_articles_repr = ReaderWriter.read_data(os.path.join(data_dir, 'processor_articles_repr.pkl'))
+    proc_config = ReaderWriter.read_data(os.path.join(data_dir, 'processor_config.pkl'))
+
+    return q_partitions, docs_data_tensor, proc_articles_dict, proc_articles_repr, proc_config
+
+
+def build_doc_partitions(questions, proc_articles_repr, proc_articles_dict, proc_config, docs_data_tensor, q_partitions,
+                         validate=False):
+    processor = DataDictProcessor(proc_config)
+    # reassign updated articles_dict and articles_repr to processor instance
+    processor.set_instance_attr(proc_articles_repr, proc_articles_dict, proc_config)
+
+    # turn the list of ids in q_partitions into dataset using PartitionDataTensor
+    q_docpartitions = {}
+    for question in questions:
+        q_docpartitions.update(generate_docpartition_per_question(docs_data_tensor, q_partitions, question))
+
+    if validate:
+        validate_q_docpartitions(q_docpartitions, q_partitions)
+
+    # add class weights in q_docpartitions
+    # we will use the weights to emphasize less occuring class during training phase
+    compute_class_weights_per_fold_(q_docpartitions)
+
+    return q_docpartitions
+
+
+def load_biobert_model(biobert_pth):
+    """Read saved state dict for biobert model on disk
+
+    Args:
+        biobert_pth: str, folder path where model state dictionary and config file are saved
+    """
+    bert_config_file = os.path.join(biobert_pth, 'bert_config.json')
+    config = BertConfig.from_json_file(bert_config_file)
+    print("Building PyTorch model from configuration: {}".format(str(config)))
+    model = BertForPreTraining(config)
+    model.load_state_dict(torch.load(os.path.join(biobert_pth, 'biobert_statedict.pkl')))
```

Comment:
Maybe we can add  `map_location=device` when loading the state_dict to have control on what gpu is loaded (although this model will be copied to whatever location on gpu later on!)

Reply:
the way the code was when you commented, was it going to gpu 0 by default? I tried to send it to gpu 1 (which I also use later for training one of the question models), but then the training process said there are no devices available. So I explicitly sent it to gpu 0 now. Is that what you were picturing?

Reply:
I assume it is like loading other pickled/saved tensors. It will load them depending on the device property (I suspect if they were on GPU1, they will be loaded on GPU1, etc.. Sending it to GPU0 is fine with me.